Add CSI driver operator docs (#25)
elmiko/ocp-infrastructure-provider-onboarding-guide: docs/container-storage-interface-driver/index.md

# Container Storage Interface Driver

## Goal

In the end, OpenShift installation should provide a cluster that is usable out of the box. Users can create PVCs or StatefulSets and get a decent storage for them.

* Install one CSI driver, the most useful for the platform. Usually it’s one that provides block volumes (AWS EBS, GCP PD, Azure Disk, OpenStack Cinder, ...).
* Provide *one* default StorageClass that is good for a generic usage. Not too expensive, but not too slow either.

>   * Each cluster is different, don’t anticipate application needs and leave creation of additional StorageClasses to the cluster administrators. They know better what they need.

Reviewer: is this advocating that the reader should not anticipate the applications needs?

if so, i would suggest this change:
```suggestion
  * Each cluster is different, don’t anticipate application needs, leave creation of additional StorageClasses to the cluster administrators. They know better what they need.
```

If possible, make the CSI driver working without metrics and add them later. They’re tricky to set up correctly. In general, follow GCP PD CSI driver operator example how the metric ports are exposed in the Pods and how their scraping is configured in [ServiceMonitor CR](https://github.com/openshift/gcp-pd-csi-driver-operator/blob/master/assets/servicemonitor.yaml).

OpenShift shall handle TLS side of things, i.e. it will provide [a Secret with TLS key](https://github.com/openshift/gcp-pd-csi-driver-operator/blob/223a251c3ba39d8af605258d14794b32a5cfafda/assets/service.yaml#L5) that can be used in [`kube-rbac-proxy` sidecars](https://github.com/openshift/gcp-pd-csi-driver-operator/blob/223a251c3ba39d8af605258d14794b32a5cfafda/assets/controller.yaml#L114-L115) using [Secret volumes](https://github.com/openshift/gcp-pd-csi-driver-operator/blob/223a251c3ba39d8af605258d14794b32a5cfafda/assets/controller.yaml#L127-L129). 

### CI

> If you do not do so yet, test your CSi driver in vanilla Kubernetes frequently using [Kubernetes storage tests](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/storage/external). No OpenShift nor operator is needed here and all your Kubernetes customers will benefit from it.

Reviewer: minor nit
```suggestion
If you do not do so yet, test your CSI driver in vanilla Kubernetes frequently using [Kubernetes storage tests](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/storage/external). Neither OpenShift or an operator is needed here and all your Kubernetes customers will benefit from it.
```

OpenShift packages the same tests as openshift-test binary / [container image](https://quay.io/repository/openshift/origin-tests). The binary just wraps the tests with proper OpenShift privileges, but otherwise they're the same tests as upstream and consume the same `manifest.yaml` file. To run the tests:

1. Place `manifest.yaml` and `kubeconfig` in a single directory, e.g.  `data/`
2. Run `openshift-tests` from `quay.io/repository/openshift/origin-tests` image:
    ```
    $ podman run -v `pwd`:/data:z --rm -it quay.io/repository/openshift/origin-tests:latest sh -c "KUBECONFIG=/data/kubeconfig TEST_CSI_DRIVER_FILES=/data/manifest.yaml /usr/bin/openshift-tests run openshift/csi --junit-dir /data/results"
    ```

In the end, a CI job will run the same tests. CI for a CSI driver is tricky to set up and probably should be done by Red Hat.

## Deliverables

### CSI driver
Fork the CSI driver under `github.com/openshift` and integrate it with Prow and Tide as described [elsewhere in this guide](../procedures).

> Make sure that the CSI driver image is built in our CI.

Reviewer: minor nit
```suggestion
Make sure that the CSI driver image is built in Red Hat's CI.
```

### openshift/api changes
Add your CSI driver name to allowed `ClusterCSIDriver` names [here](https://github.com/openshift/api/blob/b632c5fc10c0ea86cb18577db9388109a43d465f/operator/v1/types_csi_cluster_driver.go#L44-L57) and [here](https://github.com/openshift/api/blob/master/operator/v1/0000_90_cluster_csi_driver_01_config.crd.yaml-patch). You must update `cluster-storage-operator` to the new `openshift/api` version to get the `ClusterCSIDriver` CRD updated during cluster installation / update. A simple `go get -u github.com/openshift/api@master` is enough.

### CSI driver operator
The CSI driver must be installed via an operator and the operator must be based on library-go. As listed above, it has a lot of knowledge about OpenShift. **You cannot use an operator previously developed in-house and using OLM!**

> * Get a new empty repo in `gihub.com/openshift`. Take [the GCP PD CSI driver operator](https://github.com/openshift/gcp-pd-csi-driver-operator) as a base and copy it there. Replace traces of "GCP PD" and "gcp-pd" in the whole repo with your CSI driver name. If you're lucky, you may be done!

Reviewer: i think it is worth mentioning that the red hat contact will need to make the repo, and we can link to the [procedure doc](https://github.com/elmiko/ocp-infrastructure-provider-onboarding-guide/blob/main/docs/procedures/creating-an-openshift-repository.md)
